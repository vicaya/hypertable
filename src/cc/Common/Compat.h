/**
 * Copyright (C) 2007 Luke Lu (Zvents, Inc.)
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#ifndef HYPERTABLE_COMPAT_H
#define HYPERTABLE_COMPAT_H

// Include this before other includes .cc files

// The same stuff for C code
#include "compat-c.h"

#include <cstddef> // for std::size_t and std::ptrdiff_t

// C++ specific stuff
#ifndef BOOST_SPIRIT_THREADSAFE
#  define BOOST_SPIRIT_THREADSAFE
#endif

#define BOOST_IOSTREAMS_USE_DEPRECATED

#include "Sweetener.h"

#define HT_UNUSED(x) static_cast<void>(x)

#endif // HYPERTABLE_COMPAT_H
