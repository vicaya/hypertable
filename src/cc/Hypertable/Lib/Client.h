/** -*- c++ -*-
 * Copyright (C) 2010 Doug Judd (Zvents, Inc.)
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 2 of the
 * License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#ifndef HYPERTABLE_CLIENT_H
#define HYPERTABLE_CLIENT_H

#include "Common/Mutex.h"
#include "Common/ReferenceCount.h"
#include "Common/String.h"

#include "AsyncComm/ApplicationQueue.h"
#include "AsyncComm/ConnectionManager.h"
#include "Hyperspace/Session.h"

#include "MasterClient.h"
#include "NameIdMapper.h"
#include "NamespaceCache.h"
#include "TableCache.h"
#include "Namespace.h"

namespace Hypertable {

  class Comm;
  class HqlInterpreter;

  class Client : public ReferenceCount {
  public:

    /**
     * Constructs the object using the specified config file
     *
     * @param install_dir path to Hypertable installation directory
     * @param config_file name of configuration file
     * @param default_timeout_ms default method call timeout in milliseconds
     */
    Client(const String &install_dir, const String &config_file,
           uint32_t default_timeout_ms=0);

    /**
     * Constructs the object using the default config file
     *
     * @param install_dir path to Hypertable installation directory
     * @param default_timeout_ms default method call timeout in milliseconds
     */
    Client(const String &install_dir = String(), uint32_t default_timeout_ms=0);
    ~Client() {}

    /**
     * Creates a namespace
     *
     * @param name name of the namespace to create
     * @param base optional base Namespace (if specified the name parameter will be relative
     *        to base)
     * @param create_intermediate if true then create all non-existent intermediate namespaces
     */
    void create_namespace(const String &name, Namespace *base=NULL,
                          bool create_intermediate=false);

    /**
     * Opens a Namespace
     *
     * @param name name of the Namespace
     * @param base optional base Namespace (if specified the name parameter will be relative
     *        to base)
     * @return pointer to the Namespace object
     */
    NamespacePtr open_namespace(const String &name, Namespace *base=NULL);

    /**
     * Checks if the namespace exists
     *
     * @param name name of namespace
     * @param base optional base Namespace (if specified the name parameter will be relative
     *        to base)
     * @return true if namespace exists false ow
     */
    bool exists_namespace(const String &name, Namespace *base=NULL);

    /**
     * Removes a namespace.  This command instructs the Master to
     * remove a namespace from the system. The namespace must be empty, ie all tables and
     * namespaces under it must have been dropped already
     *
     * @param name namespace name
     * @param base optional base Namespace (if specified the name parameter will be relative
     *        to base)
     * @param if_exists don't throw an exception if table does not exist
     */
    void drop_namespace(const String &name, Namespace *base=NULL, bool if_exists=false);

    Hyperspace::SessionPtr& get_hyperspace_session();
    void close();
    void shutdown();

    /**
     *
     * @param immutable_namespace if true then the namespace can't be modified once its set
     */
    HqlInterpreter *create_hql_interpreter(bool immutable_namespace=true);

  private:

    void initialize();

    PropertiesPtr           m_props;
    Comm                   *m_comm;
    ConnectionManagerPtr    m_conn_manager;
    ApplicationQueuePtr     m_app_queue;
    Hyperspace::SessionPtr  m_hyperspace;
    NameIdMapperPtr         m_namemap;
    MasterClientPtr         m_master_client;
    RangeLocatorPtr         m_range_locator;
    uint32_t                m_timeout_ms;
    String                  m_install_dir;
    TableCachePtr           m_table_cache;
    NamespaceCachePtr       m_namespace_cache;
    Mutex                   m_mutex;
    bool                    m_hyperspace_reconnect;
    bool                    m_refresh_schema;
    String                  m_toplevel_dir;
  };

  typedef intrusive_ptr<Client> ClientPtr;

} // namespace Hypertable

#endif // HYPERTABLE_CLIENT_H
