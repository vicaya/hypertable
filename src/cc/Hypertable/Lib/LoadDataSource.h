/** -*- c++ -*-
 * Copyright (C) 2008 Doug Judd (Zvents, Inc.)
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 2 of the
 * License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#ifndef HYPERTABLE_LOADDATASOURCE_H
#define HYPERTABLE_LOADDATASOURCE_H

#include <fstream>
#include <iostream>
#include <string>
#include <vector>

#include <boost/iostreams/device/file.hpp>
#include <boost/iostreams/filtering_stream.hpp>

#include "Common/ByteString.h"
#include "Common/DynamicBuffer.h"
#include "Common/String.h"
#include "Common/ReferenceCount.h"

#include "DataSource.h"
#include "FixedRandomStringGenerator.h"
#include "LoadDataFlags.h"


namespace Hypertable {

  enum {
    LOCAL_FILE = 1,
    DFS_FILE,
    STDIN
  };

  class LoadDataSource : public ReferenceCount {

  public:
    LoadDataSource(const String &header_fname,
		   int row_uniquify_chars = 0,
                   int load_flags = 0);

    virtual ~LoadDataSource() { delete [] m_type_mask; return; }

    bool has_timestamps() {
      return m_leading_timestamps || (m_timestamp_index != -1);
    }

    virtual bool next(uint32_t *type_flagp, KeySpec *keyp,
                      uint8_t **valuep, uint32_t *value_lenp,
                      uint32_t *consumedp);

    virtual void init(const std::vector<String> &key_columns, const String &timestamp_column);

    int64_t get_current_lineno() { return m_cur_line; }
    unsigned long get_source_size() const { return m_source_size; }

  protected:

    bool get_next_line(String &line) {
      if (m_first_line_cached) {
	       line = m_first_line;
	       m_first_line_cached = false;
	       return true;
      }
      return getline(m_fin, line);
    }

    virtual void parse_header(const String& header,
                              const std::vector<String> &key_columns,
                              const String &timestamp_column);
    virtual void init_src()=0;
    virtual uint64_t incr_consumed()=0;

    bool should_skip(int idx, const uint32_t *masks) {
      uint32_t bm = masks[idx];
      return bm && ((bm & TIMESTAMP) ||
		    !(LoadDataFlags::duplicate_key_columns(m_load_flags) && (bm & ROW_KEY)));
    }

    class KeyComponentInfo {
    public:
      KeyComponentInfo()
          : index(0), width(0), left_justify(false), pad_character(' ') {}
      void clear() { index=0; width=0; left_justify=false; pad_character=' '; }
      int index;
      int width;
      bool left_justify;
      char pad_character;
    };

    enum TypeMask {
      ROW_KEY =           (1 << 0),
      TIMESTAMP =         (1 << 1)
    };

    String get_header();

    bool parse_date_format(const char *str, int64_t &timestamp);

    bool add_row_component(int index);

    struct ColumnInfo {
      String family;
      String qualifier;
    };

    std::vector<ColumnInfo> m_column_info;
    std::vector<const char *> m_values;
    std::vector<KeyComponentInfo> m_key_comps;
    uint32_t *m_type_mask;
    size_t m_next_value;
    boost::iostreams::filtering_istream m_fin;
    int64_t m_cur_line;
    DynamicBuffer m_line_buffer;
    DynamicBuffer m_row_key_buffer;
    bool m_hyperformat;
    bool m_leading_timestamps;
    int m_timestamp_index;
    int64_t m_timestamp;
    size_t m_limit;
    uint64_t m_offset;
    bool m_zipped;
    FixedRandomStringGenerator *m_rsgen;
    String m_header_fname;
    int m_row_uniquify_chars;
    int m_load_flags;
    String m_first_line;
    bool m_first_line_cached;
    unsigned long m_source_size;
  };

 typedef boost::intrusive_ptr<LoadDataSource> LoadDataSourcePtr;

} // namespace Hypertable

#endif // HYPERTABLE_LOADDATASOURCE_H
