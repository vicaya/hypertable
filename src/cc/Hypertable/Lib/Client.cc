/** -*- c++ -*-
 * Copyright (C) 2008 Doug Judd (Zvents, Inc.)
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 2 of the
 * License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#include "Common/Compat.h"
#include <cassert>
#include <cstdlib>

extern "C" {
#include <poll.h>
}

#include <boost/algorithm/string.hpp>

#include "AsyncComm/Comm.h"
#include "AsyncComm/ReactorFactory.h"

#include "Common/Init.h"
#include "Common/Error.h"
#include "Common/InetAddr.h"
#include "Common/Logger.h"
#include "Common/ScopeGuard.h"
#include "Common/System.h"
#include "Common/Timer.h"

#include "Hyperspace/DirEntry.h"
#include "Hypertable/Lib/Config.h"

#include "Client.h"
#include "HqlCommandInterpreter.h"

using namespace std;
using namespace Hypertable;
using namespace Hyperspace;
using namespace Config;

Client::Client(const String &install_dir, const String &config_file,
               uint32_t default_timeout_ms)
  : m_timeout_ms(default_timeout_ms), m_install_dir(install_dir) {
  ScopedRecLock lock(rec_mutex);

  if (!properties)
    init_with_policy<DefaultCommPolicy>(0, 0);

  m_props = new Properties(config_file, file_desc());
  initialize();
}

Client::Client(const String &install_dir, uint32_t default_timeout_ms)
  : m_timeout_ms(default_timeout_ms), m_install_dir(install_dir) {
  ScopedRecLock lock(rec_mutex);

  if (!properties)
    init_with_policy<DefaultCommPolicy>(0, 0);

  if (m_install_dir.empty())
    m_install_dir = System::install_dir;

  m_props = properties;
  initialize();
}

void Client::create_namespace(const String &name, Namespace *base, bool create_intermediate) {

  String full_name;
  String sub_name = name;
  int flags=NameIdMapper::IS_NAMESPACE;

  if(create_intermediate)
    flags |= NameIdMapper::CREATE_INTERMEDIATE;

  if (base != NULL) {
    full_name = base->get_name() + '/';
  }
  full_name += sub_name;

  Namespace::canonicalize(&full_name);
  m_master_client->create_namespace(full_name, flags);
}

NamespacePtr Client::open_namespace(const String &name, Namespace *base) {
  String full_name;
  String sub_name = name;

  Namespace::canonicalize(&sub_name);

  if (base != NULL) {
    full_name = base->get_name() + '/';
  }

  full_name += sub_name;

  return m_namespace_cache->get(full_name);
}

bool Client::exists_namespace(const String &name, Namespace *base) {
  String id;
  bool is_namespace = false;
  String full_name;
  String sub_name = name;

  Namespace::canonicalize(&sub_name);

  if (base != NULL) {
    full_name = base->get_name() + '/';
  }

  full_name += sub_name;

  if (!m_namemap->name_to_id(full_name, id, &is_namespace) || !is_namespace)
    return false;

  // TODO: issue 11

  String namespace_file = m_toplevel_dir + "/tables/" + id;

  try {
    return m_hyperspace->exists(namespace_file);
  }
  catch(Exception &e) {
    HT_THROW(e.code(), e.what());
  }
}

void Client::drop_namespace(const String &name, Namespace *base, bool if_exists) {
  String full_name;
  String sub_name = name;

  Namespace::canonicalize(&sub_name);

  if (base != NULL) {
    full_name = base->get_name() + '/';
  }

  full_name += sub_name;

  m_namespace_cache->remove(full_name);
  m_master_client->drop_namespace(full_name, if_exists);
}

Hyperspace::SessionPtr& Client::get_hyperspace_session()
{
  return m_hyperspace;
}

void Client::close() {
  HT_WARN("close() is no longer supported");
}

void Client::shutdown() {
  m_master_client->shutdown();
}


HqlInterpreter *Client::create_hql_interpreter(bool immutable_namespace) {
  return new HqlInterpreter(this, this->m_conn_manager, immutable_namespace);
}

// ------------- PRIVATE METHODS -----------------
void Client::initialize() {
  uint32_t wait_time, remaining;
  uint32_t interval=5000;

  m_toplevel_dir = m_props->get_str("Hypertable.Directory");
  boost::trim_if(m_toplevel_dir, boost::is_any_of("/"));
  m_toplevel_dir = String("/") + m_toplevel_dir;

  m_comm = Comm::instance();
  m_conn_manager = new ConnectionManager(m_comm);

  if (m_timeout_ms == 0)
    m_timeout_ms = m_props->get_i32("Hypertable.Request.Timeout");

  m_hyperspace_reconnect = m_props->get_bool("Hyperspace.Session.Reconnect");

  m_hyperspace = new Hyperspace::Session(m_comm, m_props);

  m_namemap = new NameIdMapper(m_hyperspace, m_toplevel_dir);

  m_refresh_schema = m_props->get_bool("Hypertable.Client.RefreshSchema");

  Timer timer(m_timeout_ms, true);

  remaining = timer.remaining();
  wait_time = (remaining < interval) ? remaining : interval;

  while (!m_hyperspace->wait_for_connection(wait_time)) {

    if (timer.expired())
      HT_THROW_(Error::CONNECT_ERROR_HYPERSPACE);

    cout << "Waiting for connection to Hyperspace..." << endl;

    remaining = timer.remaining();
    wait_time = (remaining < interval) ? remaining : interval;
  }

  m_app_queue = new ApplicationQueue(m_props->
                                     get_i32("Hypertable.Client.Workers"));
  m_master_client = new MasterClient(m_conn_manager, m_hyperspace, m_toplevel_dir,
                                     m_timeout_ms, m_app_queue);

  m_master_client->initiate_connection();

  if (!m_master_client->wait_for_connection(timer))
    HT_THROW(Error::REQUEST_TIMEOUT, "Waiting for Master connection");

  m_range_locator = new RangeLocator(m_props, m_conn_manager, m_hyperspace,
                                     m_timeout_ms);
  m_table_cache = new  TableCache(m_props, m_range_locator, m_conn_manager, m_hyperspace,
                                  m_app_queue, m_namemap, m_timeout_ms);
  m_namespace_cache = new NamespaceCache(m_props, m_range_locator, m_conn_manager, m_hyperspace,
                                         m_app_queue, m_namemap, m_master_client,
                                         m_table_cache, m_timeout_ms);
}
