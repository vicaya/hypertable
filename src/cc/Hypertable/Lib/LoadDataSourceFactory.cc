/** -*- c++ -*-
 * Copyright (C) 2009 Sanjit Jhala (Zvents, Inc.)
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 2 of the
 * License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#include "Common/Compat.h"
#include "HqlInterpreter.h"
#include "LoadDataSourceFactory.h"
#include "LoadDataSourceFileLocal.h"
#include "LoadDataSourceFileDfs.h"
#include "LoadDataSourceStdin.h"

using namespace Hypertable;
using namespace std;

/**
 *
 */
LoadDataSource *
LoadDataSourceFactory::create(DfsBroker::ClientPtr &dfs_client,
    const String &input_fname, const int src,
    const String &header_fname, const int header_src,
    const std::vector<String> &key_columns, const String &timestamp_column,
    int row_uniquify_chars, int load_flags) {

  LoadDataSource *lds;

  switch (src) {
    case LOCAL_FILE:
      lds = new LoadDataSourceFileLocal(input_fname, header_fname,
                                        row_uniquify_chars, load_flags);
      break;
    case DFS_FILE:
      lds = new LoadDataSourceFileDfs(dfs_client, input_fname, header_fname,
                                      row_uniquify_chars, load_flags );
      break;
    case STDIN:
      lds = new LoadDataSourceStdin(header_fname, row_uniquify_chars, load_flags);
      break;
    default:
      HT_THROW(Error::HQL_PARSE_ERROR, "LOAD DATA - bad filename");
  }

  lds->init(key_columns, timestamp_column);
  return lds;
}

