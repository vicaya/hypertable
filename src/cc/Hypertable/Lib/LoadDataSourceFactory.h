/** -*- c++ -*-
 * Copyright (C) 2009 Sanjit Jhala (Zvents, Inc.)
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 2 of the
 * License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#ifndef HYPERTABLE_LOADDATASOURCEFACTORY_H
#define HYPERTABLE_LOADDATASOURCEFACTORY_H

#include "Common/String.h"

#include "LoadDataSource.h"
#include "DfsBroker/Lib/Client.h"

namespace Hypertable {

  class LoadDataSourceFactory {

  public:

    static LoadDataSource *create(DfsBroker::ClientPtr &dfs_client,
                                  const String &fname, const int src,
                                  const String &header_fname, const int header_src,
                                  const std::vector<String> &key_columns,
                                  const String &timestamp_column,
                                  int row_uniquify_chars = 0,
                                  int load_flags = 0
                                  );
  };

} // namespace Hypertable

#endif // HYPERTABLE_LOADDATASOURCEFACTORY_H
