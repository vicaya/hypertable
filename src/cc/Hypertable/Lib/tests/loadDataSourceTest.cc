/** -*- c++ -*-
 * Copyright (C) 2008 Doug Judd (Zvents, Inc.)
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; version 2 of the
 * License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#include "Common/Compat.h"
#include <cstdlib>
#include <iostream>
#include <vector>

extern "C" {
#include <errno.h>
#include <fcntl.h>
}

#include "Common/String.h"

#include "Hypertable/Lib/KeySpec.h"
#include "Hypertable/Lib/LoadDataSource.h"
#include "Hypertable/Lib/LoadDataSourceFactory.h"
#include "DfsBroker/Lib/Client.h"

using namespace Hypertable;
using namespace std;

int main(int argc, char **argv) {
  LoadDataSourcePtr lds;
  KeySpec key;
  uint8_t *value;
  uint32_t value_len;
  int fd;
  std::vector<String> key_columns;
  DfsBroker::ClientPtr null_dfs_client;

  vector<String> testnames;
  testnames.push_back("loadDataSourceTest");
  testnames.push_back("loadDataSourceTest-header");
  testnames.push_back("loadDataSourceTest-qualified-header");

  for(size_t i = 0; i < testnames.size(); i++) {
    String output_fn = testnames[i] + ".output";

    if ((fd = open(output_fn.c_str(), O_WRONLY|O_CREAT|O_TRUNC, 0644)) < 0) {
      perror("open");
      return 1;
    }

    close(2);
    dup(fd);

    key_columns.clear();
    String dat_fn = testnames[i] + ".dat";
    lds = LoadDataSourceFactory::create(null_dfs_client,
                                        dat_fn.c_str(), LOCAL_FILE, "", LOCAL_FILE,
                                        key_columns, "");

    while (lds->next(0, &key, &value, &value_len, 0)) {
      cerr << "row=" << (const char *)key.row
           << " column_family=" << key.column_family;
      if (key.column_qualifier_len > 0)
        cerr << " column_qualifier=" << (const char *)key.column_qualifier;
      cerr << " value=" << (const char *)value << endl;
    }

    String golden_fn = testnames[i] + ".golden";
    String sys_cmd = "diff " + output_fn + " " + golden_fn;
    if (system(sys_cmd.c_str()) != 0)
      return 1;
  }

  return 0;
}
