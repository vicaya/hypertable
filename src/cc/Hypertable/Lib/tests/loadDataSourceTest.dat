#timestamp	rowkey	columnkey	value
2009-02-12 02:12:12.1000000000	abenteric	foo:http://www.foo.com/	How now brown cow
2009-02-12 02:12:12:2	barothermohygrograph	bar	Herbie Hancock
2009-02-12 02:12:12	Batonga	foo:	Head Hunters
2009-02-12 02:12:12
abracadabra	Batonga	foo:	Head Hunters
2009-02-12 02:12:12	barothermohygrogra
2009-02-12 02:12:12	barothermohygrograph	ba
2009-02-12 02:12:12.1000000000	carbonification	sherlock:holmes	this\tis\ta\ttab\tdelimited\tvalue
2009-02-12 02:12:12.3000000000	biomagnify	sherlock:holmes	\\\this\contains\lots\of\backslashes\\\
