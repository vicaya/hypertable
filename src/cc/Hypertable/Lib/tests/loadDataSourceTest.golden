row=abenteric column_family=foo column_qualifier=http://www.foo.com/ value=How now brown cow
row=barothermohygrograph column_family=bar value=Herbie Hancock
row=Batonga column_family=foo value=Head Hunters
warning: too few fields on line 5
warn: invalid timestamp format on line 6, skipping...
warning: too few fields on line 7
warning: too few fields on line 8
row=carbonification column_family=sherlock column_qualifier=holmes value=this\tis\ta\ttab\tdelimited\tvalue
row=biomagnify column_family=sherlock column_qualifier=holmes value=\\\this\contains\lots\of\backslashes\\\
